[docs] Document sharing makeStyles between components (#15234)
mui-org/material-ui: docs/src/pages/css-in-js/advanced/advanced.md

## CSS injection order

> It's **really important** to understand how the CSS specificity is calculated by the browser.

> It's one of the key element to know when overriding styles.

Reviewer: ```suggestion
It's one of the key elements to know when overriding styles.
```

We **encourage** you to read this MDN paragraph: [How is specificity calculated?](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity#How_is_specificity_calculated)

By default, the style tags are injected **last** in the `<head>` element of the page.
They gain more specificity than any other style tags on your page e.g. CSS modules, styled components.

### injectFirst

> The `StylesProvider` component has a `injectFirst` prop to inject the style tags **first** in the head (less priority):

Reviewer: ```suggestion
The `StylesProvider` component has an `injectFirst` prop to inject the style tags **first** in the head (less priority):
```

</StylesProvider>
```

### makeStyles / withStyles / styled

> The injection of style tags happens in the **same order** as the makeStyles / withStyles / styled invocations. For instance the color red wins in this case:

Reviewer: ```suggestion
The injection of style tags happen in the **same order** as the makeStyles / withStyles / styled invocations. For instance the color red wins in this case:
```

Reviewer: This is great information to add to the docs 🚀

Author: The injection "happens" right, not the style tags, so I believe the `s` should remain.

Reviewer: Happens is the verb form. You are correct :+1:

export default App;
```

### makeStyles

> Whenever you need to make use of the `useStyles` hook, you should instantiate a new instance by calling `makeStyles()` in each and every module. If you import an instance from another module, the CSS injection order could be broken. It may also only manifest when building in `production` mode, so be careful with that if you witness different behavior between `dev` and `production`.

Reviewer: ```suggestion
When using `useStyles`, you should instantiate a new instance by calling `makeStyles()` in each file. If you import an instance from another file, the CSS injection order can break and cause different behaviour in `dev` and `production`.
```
What do you think of this instead? I would get another contributors opinion too.

Author: That’s indeed way cleaner, thx!
